pennylane/transforms/compile.py

```diff
@@ -0,0 +1,149 @@
+
+    Visually, the original function looks like this:
+
+    >>> dev = qml.device('default.qubit', wires=[0, 1, 2])
+    >>> qnode = qml.QNode(qfunc, dev)
+    >>> print(qml.draw(qnode)(0.2, 0.3, 0.4))
+     0: ──H──RX(0.4)──────╭X─────────RX(0.2)──╭X───────┤ ⟨Z⟩
+     1: ──H───────────╭X──╰C──────────────────╰C──╭CY──┤
+     2: ──H──RZ(0.4)──╰C───RZ(-0.4)──RX(0.3)───Y──╰CY──┤
+
+    We can compile it down to a smaller set of gates using the ``qml.compile``
+    transform.
+
+    >>> compiled_qfunc = qml.compile()(qfunc)
+    >>> compiled_qnode = qml.QNode(compiled_qfunc, dev)
+    >>> print(qml.draw(compiled_qnode)(0.2, 0.3, 0.4))
+     0: ──H───RX(0.6)───────────────────┤ ⟨Z⟩
+     1: ──H──╭X─────────────────╭CY─────┤
+     2: ──H──╰C────────RX(0.3)──╰CY──Y──┤
+
+    You can change up the set of transforms by passing a custom ``pipeline`` to
+    ``qml.compile``. The pipeline is a list of transform functions. Furthermore,
+    you can specify a number of passes (repetitions of the pipeline), and a list
+    of gates into which the compiler will first attempt to decompose the
+    existing operations prior to applying any optimization transforms.
+
+    .. code-block:: python3
+
+        compiled_qfunc = qml.compile(
+            pipeline=[
+                qml.transforms.commute_controlled(direction="left"),
+                qml.transforms.merge_rotations(atol=1e-6),
+                qml.transforms.cancel_inverses
+            ],
+            basis_set=["CNOT", "RX", "RY", "RZ"],
+            num_passes=2
+        )(qfunc)
+
+        compiled_qnode = qml.QNode(compiled_qfunc, dev)
+
+    >>> print(qml.draw(compiled_qnode)(0.2, 0.3, 0.4))
+     0: ──RZ(1.57)──RX(1.57)──RZ(1.57)───RX(0.6)───────────────────────────────────────────────────────────────────────┤ ⟨Z⟩
+     1: ──RZ(1.57)──RX(1.57)──RZ(1.57)──╭X────────RZ(1.57)──────────────────────────────────────────╭C─────────────╭C──┤
+     2: ──RZ(1.57)──RX(1.57)──RZ(1.57)──╰C────────RX(0.3)───RZ(1.57)──RY(3.14)──RZ(1.57)──RY(1.57)──╰X──RY(-1.57)──╰X──┤
+
+
+    """
+    if pipeline is None:
+        pipeline = default_pipeline
+
+    # Expand the tape; this is done to unroll any templates that may be present,
+    # as well as to decompose over a specified basis set
+    # First, though, we have to stop whatever tape may be recording so that we
+    # don't queue anything as a result of the expansion or transform pipeline
+    current_tape = get_active_tape()
+
+    with current_tape.stop_recording():
+        if basis_set is not None:
+            expanded_tape = tape.expand(depth=5, stop_at=lambda obj: obj.name in basis_set)
+        else:
+            # Expand only the templates
+            expanded_tape = tape.expand(stop_at=lambda obj: obj.name not in template_list)
+
+        # Apply the compilation transforms
+        for _ in range(num_passes):
+            for transform in pipeline:
+                expanded_tape = transform.tape_fn(expanded_tape)
```

Comment:
could this be reduced to?

```python
if basis_set is not None:
    expanded_tape = qml.invisible(tape.expand)(depth=5, stop_at=lambda obj: obj.name in basis_set)
else:
    # Expand only the templates
    expanded_tape = qml.invisible(tape.expand)(stop_at=lambda obj: obj.name not in template_list)

# Apply the compilation transforms
for _ in range(num_passes):
    for transform in pipeline:
        expanded_tape = qml.invisible(transform.tape_fn)(expanded_tape)
```

Although, this might be slightly slower...

Reply:
Definitely, just thinking it is clearer to see `stop_recording` rather than `invisible` as well.